// mvc/lib/abstract_controller.js
var stringTools = require('./utils/string'),
    formidable = require('formidable'),
    View = require('./view');

var AbstractController = function(intent){
	this._initAbstractController(intent);
};

AbstractController.prototype = {
	_initAbstractController: function(intent){
		this.renderScriptStatus = true;
		this.request = intent.request;
		this.response = intent.response;
		this.params = intent.params;
		this.controllerName = intent.controllerName;
		this.actionName = stringTools.dashToCamel(intent.actionName) + 'Action';
		this.view = new View(intent);
	},
	run: function(){
		if(this[this.actionName]){
			this[this.actionName]();
			this.renderScriptStatus && this.render();
		}else{
			res.writeHead(404, {'Content-Type': 'text/html; charset=utf-8'});
			res.end("<h1>Controller NOT Found!</h1><hr />Powered by dk");
		}
	},
	render: function (){
		var self = this;
		this.view.renderScript(function(viewContent){
			self.response.writeHead(200, {'Content-Type': 'text/html; charset=utf-8'});
			self.response.end(viewContent);
		});
	},

	json: function(content){
		this.response.writeHead(200, {'Content-Type': 'application/json;charset=utf-8'})
		this.response.end(content);
	},

	setNoRender: function(){
		this.renderScriptStatus = false;
	},
    /**
     * 设置Header
     * @param name
     * @param value
     */
    setHeader: function(name, value){
        this.response.setHeader(name, value);
    },
    /**
     * 获取原始POST数据
     * @param callback(error, data)
     */
    getRawPost: function(callback){
        if(this.request.method == "POST"){
            var postData = '';
            this.request.addListener('data', function(chunk){
                postData += chunk;
            });

            this.request.addListener('end', function(){
                callback(null, postData);
            });
        }else{
            callback('NOT POST', null);
        }
    },
    /**
     * 获取表单数据
     * @param callback
     */
    getPost: function(callback){
        if(this.request.method == "POST"){
            var form = new formidable.IncomingForm();
            form.parse(this.request, function(err, fields, files) {
                callback(err, fields, files);
            });
        }else{
            callback('NOT POST', null, null);
        }

    }
};

module.exports = AbstractController;
