// mvc/lib/config.js
var oo = require('./utils/oo');

var config = {};
config.serverinfo = {
    port: 8001
};

config.db = {
    host: '',
    user: '',
    password: '',
    database: '',
    charset: ''
};

config.cache = {
	host: '127.0.0.1',
	port: 6379
};

config.template = {
    isCache: false
};

config.app_path = '';
config.file_path = '';


config.routes = [
    {
        url: '/{controller}/{action}',
        module: 'api',
        controller: 'index',
        action: 'index'
    }
];

config.addConfig = function(configInstance){
    for(var key in configInstance){
        if(configInstance[key] instanceof Array && config[key]){
            config[key] = configInstance[key].concat(config[key]);
            continue;
        }

        if(configInstance[key].constructor == String){
            config[key] = configInstance[key];
            continue;
        }

        if(configInstance[key] instanceof Object){
            config[key] = oo.mix(config[key], configInstance[key]);
            continue;
        }
    }
}

module.exports = config;
